src/dataaggregator/aggregator/policyPoller_test.go

```diff
@@ -0,0 +1,96 @@
+package aggregator_test
+
+import (
+	"code.cloudfoundry.org/lager"
+	. "dataaggregator/aggregator"
+	"dataaggregator/aggregator/fakes"
+	. "dataaggregator/policy"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"time"
+)
+
+var _ = Describe("PolicyPoller", func() {
+	var (
+		database *fakes.FakeDB
+
+		poller     *PolicyPoller
+		testAppId1 = "testAppId"
+		policyStr1 = `
+		{
+		   "instance_min_count":1,
+		   "instance_max_count":5,
+		   "scaling_rules":[
+		      {
+		         "metric_type":"MemoryUsage",
+		         "stat_window_secs":300,
+		         "breach_duration_secs":300,
+		         "threshold":30,
+		         "operator":"<",
+		         "cool_down_secs":300,
+		         "adjustment":"-1"
+		      }
+		   ]
+		}`
+	)
+
+	BeforeEach(func() {
+		database = &fakes.FakeDB{}
+		logger := lager.NewLogger("PolicyPoller-test")
+		poller = NewPolicyPoller(logger, TestPolicyPollerInterval, database)
+	})
+	Context("Start", func() {
+		BeforeEach(func() {
+			poller.Start()
+		})
+		AfterEach(func() {
+			poller.Stop()
+		})
+		Context("when the poller is started", func() {
+			BeforeEach(func() {
+				database.RetrievePoliciesStub = func() ([]*PolicyJson, error) {
+					return []*PolicyJson{&PolicyJson{AppId: testAppId1, PolicyStr: policyStr1}}, nil
+				}
+			})
+			It("should retrieve policies for every interval", func() {
+				Eventually(database.RetrievePoliciesCallCount, 2500*time.Millisecond, 500*time.Millisecond).Should(Equal(2))
+			})
+		})
+		Context("when there is no policy", func() {
+			BeforeEach(func() {
+				database.RetrievePoliciesStub = func() ([]*PolicyJson, error) {
+					return []*PolicyJson{}, nil
+				}
+			})
+			It("no trigger generated", func() {
+				Consistently(func() int { return len(poller.PolicyMap) }, 2500*time.Millisecond, 500*time.Millisecond).Should(Equal(0))
```

Comment:
You need to examine the stacks between the WRITE and READ. In most cases it is more a matter of adding sync blocks. I had you add -race because I could see the issue in how you are testing which is why I asked how the Poller is going to be used.

I would personally break this apart into 2 pieces.
1. A poller that calls some operation on a periodic basis that can be stopped.
2. The ability to retrieve policies
3. There is something that is going to be called after retrieving policies which is passed the new set of policies

It becomes easier to test and by having these additional entry points, you can do the testing on the right goroutines.